cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,378 @@
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    using strings::detail::create_string_vector_from_column;
+    auto const source_vector = create_string_vector_from_column(source, stream);
+    auto const begin = source_vector.begin();
+    auto const end = begin + scatter_rows;
+    return strings::detail::scatter(begin, end, scatter_iter, target, mr, stream);
+  }
+};
+
+struct scatter_impl {
+  template <typename T, std::enable_if_t<std::is_integral<T>::value
+     and not std::is_same<T, bool8>::value>* = nullptr>
+  std::unique_ptr<table> operator()(
+      table_view const& source, column_view const& scatter_map,
+      table_view const& target, bool check_bounds,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    if (check_bounds) {
+      auto const begin = -target.num_rows();
+      auto const end = target.num_rows();
+      auto bounds = bounds_checker<T>{begin, end};
+      CUDF_EXPECTS(thrust::all_of(rmm::exec_policy(stream)->on(stream),
+        scatter_map.begin<T>(), scatter_map.end<T>(), bounds),
+        "Scatter map index out of bounds");
+    }
+
+    // Transform negative indices to index + target size
+    auto scatter_rows = scatter_map.size();
+    auto scatter_iter = thrust::make_transform_iterator(
+      scatter_map.begin<T>(), index_converter<T>{target.num_rows()});
+
+    // Second dispatch over data type per column
+    auto result = std::vector<std::unique_ptr<column>>(target.num_columns());
+    auto scatter_functor = column_scatterer<decltype(scatter_iter)>{};
+    std::transform(source.begin(), source.end(), target.begin(), result.begin(),
+      [=](auto const& source_col, auto const& target_col) {
+        return type_dispatcher(source_col.type(), scatter_functor,
+          source_col, scatter_iter, scatter_rows, target_col, mr, stream);
+      });
+
+    auto gather_map = make_gather_map<T>(scatter_iter, scatter_rows,
+      target.num_rows(), stream);
+    gather_bitmask(source, gather_map.begin(), result, mr, stream);
+
+    return std::make_unique<table>(std::move(result));
+  }
+
+  template <typename T, std::enable_if_t<not std::is_integral<T>::value
+      or std::is_same<T, bool8>::value>* = nullptr>
+  std::unique_ptr<table> operator()(
+      table_view const& source, column_view const& scatter_map,
+      table_view const& target, bool check_bounds,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    CUDF_FAIL("Scatter map column must be an integral, non-boolean type");
+  }
+};
+
+template <bool mark_true, typename MapIterator>
+__global__ void marking_bitmask_kernel(
+    mutable_column_device_view destination,
+    MapIterator scatter_map,
+    size_type num_scatter_rows)
+{
+  size_type row = threadIdx.x + blockIdx.x * blockDim.x;
+
+  while (row < num_scatter_rows) {
+    size_type const output_row = scatter_map[row];
+
+    if (mark_true){
+      destination.set_valid(output_row);
```

Comment:
Yikes, this is going to be slow. I thought the `gather_bitmask_kernel` was being used?

Reply:
It's being used for the regular scatter, but it needed significant changes to make it work with the scalar variant (no source bitmask, just a single valid flag). I just ported the existing scatter scalar as-is, even though it did look like it could be improved. The changes @cwharris is making for gather_bitmask_kernel to take a selector for the source validity could fix this